src/neuroglancer/sliceview/compresso/index.ts

```diff
@@ -0,0 +1,129 @@
+/**
+ * @license
+ * Copyright William Silvermsith
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {TypedArray} from 'neuroglancer/util/array';
+
+// @ts-ignore
+import createCompressoModule from './compresso.js';
+
+const compressoModulePromise : any = createCompressoModule();
+
+// not a full implementation of read header, just the parts we need
+function read_header(buffer: Uint8Array) : Map<string, number> {
+  // check for header "cpso"
+  const magic = (
+       buffer[0] === 'c'.charCodeAt(0) && buffer[1] === 'p'.charCodeAt(0)
+    && buffer[2] === 's'.charCodeAt(0) && buffer[3] === 'o'.charCodeAt(0)
+  );
+  if (!magic) {
+    throw new Error("compresso: didn't match magic numbers")
+  }
+  const format = buffer[4];
+  if (format !== 0) {
+    throw new Error("compresso: didn't match format version")
+  }
+
+  let u16 = (lower : number, upper: number) => { return (lower|0) + (upper << 8) };
+
+  const data_width = buffer[5];
+  const sx = u16(buffer[6], buffer[7]);
+  const sy = u16(buffer[8], buffer[9]);
+  const sz = u16(buffer[10], buffer[11]);
+
+  let map = new Map<string,number>();
+  map.set("sx", sx);
+  map.set("sy", sy);
+  map.set("sz", sz);
+  map.set("data_width", data_width);
+
+  return map;
+}
+
+function read_image(
+  m:any, image_ptr:number,  
+  sx:number, sy:number, sz:number, data_width:number
+) : TypedArray {
+
+  const voxels = sx * sy * sz;
+  let image;
+
+  let memory = m.HEAPU8.buffer;
+
+  if (data_width === 1) {
+    image = new Uint8Array(memory, image_ptr, voxels);
+  }
+  else if (data_width === 2) {
+    image = new Uint16Array(memory, image_ptr, voxels);
+  }
+  else if (data_width === 4) {
+    image = new Uint32Array(memory, image_ptr, voxels);
+  }
+  else if (data_width === 8) {
+    image = new Uint32Array(memory, image_ptr, voxels * 2);
+  }
+  else {
+    throw new Error(`data_width must be 1, 2, 4, or 8. Got: ${data_width}`);
+  }
+
+  // copy the array so it can be memory managed by JS
+  // and we can free the emscripten buffer
+  return image.slice(0); 
+}
+
+export function decompress_compresso(buffer: Uint8Array) 
```

Comment:
Could make this function async and then use `await` rather than `then`.

Reply:
I tried doing this but I saw a large slowdown. I think the module was getting reinitialized. Typescript didn't like a top-level await. Could you share a snippet? 
```
src/neuroglancer/sliceview/compresso/index.ts:20:37: error: Top-level await is not available in the configured target environment ("ES2017") Error: Build failed with 1 error:
```

Reply:
Sorry, I meant just for decompress_compresso, not for compressoModulePromise.  E.g.

```typescript
export async function decompressCompresso(buffer: Uint8Array): Promise<TypedArray> {
  const m = await compressoModulePromise;
  const header = ...;
  ...
}
```

Reply:
Huh. I'm a bit hazy on why I can use a return and not a yield and with what happens to the promise, but that works. I'll have to read up some more on this topic.

Reply:
Ok I read up. I get it now!